HoRNDIS.cpp

```diff
@@ -968,120 +941,132 @@ UInt32 HoRNDIS::outputPacket(mbuf_t packet, void *param) {
 	mbuf_copydata(packet, 0, pktlen, hdr + 1);
 	
 	freePacket(packet);
+	packet = NULL;
 	
-	/* Now, fire it off! */
-	outbufs[poolIndx].comp.target    = this;
-	outbufs[poolIndx].comp.parameter = (void *)poolIndx;
-	outbufs[poolIndx].comp.action    = dataWriteComplete;
+	// Now, fire it off!
+	IOUSBHostCompletion *const comp = &outbufs[poolIndx].comp;
+	comp->owner     = this;
+	comp->parameter = (void *)(uintptr_t)poolIndx;
+	comp->action    = dataWriteComplete;
 	
-	ior = fOutPipe->Write(outbufs[poolIndx].mdp, &outbufs[poolIndx].comp);
+	ior = robustIO(fOutPipe, &outbufs[poolIndx], transmitLength);
 	if (ior != kIOReturnSuccess) {
-		LOG(V_ERROR, "write failed");
-		if (ior == kIOUSBPipeStalled) {
-			fOutPipe->Reset();
-			ior = fOutPipe->Write(outbufs[poolIndx].mdp, &outbufs[poolIndx].comp);
-			if (ior != kIOReturnSuccess) {
-				LOG(V_ERROR, "write really failed");
-				fpNetStats->outputErrors++;
-				return ior;
-			}
+		if (isTransferStopStatus(ior)) {
+			LOG(V_DEBUG, "WRITER: The device was possibly disconnected: ignoring the error");
+		} else {
+			LOG(V_ERROR, "write failed: %08x", ior);
+			fpNetStats->outputErrors++;
 		}
+		// Packet was already freed: just quit:
+		return kIOReturnOutputDropped;
 	}
+	// Only here - when 'fOutPipe->io' has fired - we mark the buffer in-use:
+	numFreeOutBufs--;
+	fCallbackCount++;
 	fpNetStats->outputPackets++;
-	
-	return kIOReturnOutputSuccess;
+	// If we ran out of free buffers, issue a stall command to the queue.
+	// Note, this would be "we accept this packet, but don't give us more yet",
+	// which is NOT the same as 'kIOReturnOutputStall'.
+	const bool stallQueue = (numFreeOutBufs == 0);
+	if (stallQueue) {
+		LOG(V_PACKET, "Issuing stall command to the output queue");
```

Comment:
Maybe V_NOTE might be better ... this really shouldn't happen

Reply:
Err... Well, this should and WILL happen pretty often. E.g. if you go to "speedtest.net" and run the test. Please keep in mind that we have a host that can push the network data at a high rate, and a relatively slow USB-2 interface and possibly Cellular connection. So, there must be a flow-control mechanism, and it's happening right here, by stalling/resuming the queue.

Reply:
Interesting.  I thought people just dropped packets in this case.